contracts/NFTVesting/IntervalVestingNFTTimeLock.sol

```diff
@@ -0,0 +1,169 @@
+        );
+        _nft = nft_;
+        _tokenId = tokenId_;
+        _beneficiary = beneficiary_;
+        _releaseTime = releaseTime_;
+        _maxDiscount = maxDiscount_;
+        _maxInterval = maxInterval_;
+        _intervalDuration = intervalDuration_;
+        _currentInterval = 0;
+    }
+
+    /**
+     * @dev Returns the smart contract NFT.
+     */
+    function nft() public view virtual returns (IERC721) {
+        return _nft;
+    }
+
+    /**
+     * @dev Returns the token ID of the NFT being held.
+     */
+    function tokenId() public view virtual returns (uint256) {
+        return _tokenId;
+    }
+
+    /**
+     * @dev Returns the beneficiary that will receive the NFT.
+     */
+    function beneficiary() public view virtual returns (address) {
+        return _beneficiary;
+    }
+
+    /**
+     * @dev Returns the time when the NFT are released in seconds since Unix Interval (i.e. Unix timestamp).
+     */
+    function releaseTime() public view virtual returns (uint256) {
+        return _releaseTime;
+    }
+
+    /**
+     * @dev Returns discount for locking at each Interval
+     */
+    function discountPerInterval() public view virtual returns (uint256) {
+        return _maxDiscount / _maxInterval;
+    }
+
+    /**
+     * @dev Returns current Interval
+     */
+    function getCurrentInterval() public view returns (uint256) {
+        if (block.timestamp < _releaseTime) {
+            return 0;
+        }
+        uint256 interval = uint256(block.timestamp - _releaseTime) /
+            _intervalDuration;
+        if (interval > _maxInterval) {
+            interval = _maxInterval;
+        }
+        return interval;
+    }
+
+    /**
+     * @dev Returns the remaining interval before max vesting interval
+     */
+    function getIntervalsLeft() public view returns (uint256) {
+        return _maxInterval - _currentInterval;
+    }
+
+    /**
+     * @dev Updates the interval after release schedule
+     */
+    function updateInterval() private validRelease {
+        require(_currentInterval <= _maxInterval, "Vesting Schedule is over.");
+        // floor the current Interval to the nearest Interval interval
+        _currentInterval = getCurrentInterval();
+    }
+
+    /**
+     * @dev Get the current discount in terms of percentage for the vesting schedule.
+     */
+    function getDiscount() public view returns (uint256) {
+        require(_currentInterval > 0, "Vesting Schedule is not Up yet.");
+        return (_currentInterval * discountPerInterval()) / 100;
+    }
+
+    /**
+     * @dev Transfers NFT held by the timelock to the beneficiary. Will only succeed if invoked after the release
+     * time. Sends the discount in Eth to the beneficiary.
+     */
+
+    function release() public virtual validRelease {
+        require(
+            nft().ownerOf(tokenId()) == address(this),
+            "TimeLock: no NFT to release for user"
+        );
+
+        updateInterval();
```

Comment:
This seems to be more complicated than necessary

How about having instead:

- discount per interval
- max intervals
- interval duration

And having a requirement to check that the initialisation has ETH = max intervals * discount per interval

Such that we don't need to updateInterval which opens up unknown security implications

Instead, we could then do (in psuedo code) in discount:

vested_time = block.timestamp - _releaseTime
intervals_passed = maximum(quotient of vested_time / interval_duration, max_intervals)
discount = intervals_passed * discount_per_interval

Reply:
good idea, updateInterval feels kind of weak and suspicious.

Reply:
From what i have searched, I am not sure if its possible for developer to send eth on initialization, therefore i think ill put the check for `ETH = max intervals * discount per interval` on release